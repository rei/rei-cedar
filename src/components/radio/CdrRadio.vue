<script>
  export default {
    inheritAttrs: false,
    customOptions: {}
  }
</script>
<script setup>
import { useCssModule, computed, ref } from 'vue';

import propValidator from '../../utils/propValidator.js';
import CdrLabelWrapper from '../labelWrapper/CdrLabelWrapper.vue';
import sizeProps from '../../props/size.js';
import backgroundProps from '../../props/background.js';

const props = defineProps({
    /**
     * Class that is added to the label for custom styles
     */
    labelClass: String,
    /**
     * Class that is added to the input for custom styles
     */
    inputClass: String,
    /**
     * Class that is added to the slot wrapper for custom styles
     */
    contentClass: String,
    /**
     * Sets the name of the radio. Required.
    */
    name: {
      type: String,
      required: true,
    },

    modifier: {
      type: String,
      default: '',
      validator: (value) => propValidator(value, ['', 'hide-figure']),
    },

    size: sizeProps,

    // Set which background type the input renders on
    background: backgroundProps,

    /**
     * Sets the value of the radio. Required.
    */
    customValue: {
      type: [String, Number, Boolean, Object, Array, Symbol, Function],
    },

    /** @ignore v-model binding */
    modelValue: {
      type: [String, Number, Boolean, Object, Array, Symbol, Function],
    },
})
const emit = defineEmits(['update:modelValue']);

const baseClass = 'cdr-radio';
const newValue = ref(props.modelValue);
const radioModel = computed({
  get() {
    return props.modelValue
  },
  set(newValue) {
    emit('update:modelValue', newValue)
  }
})
const style = useCssModule();
</script>

<template>
  <cdr-label-wrapper
    :class="style[baseClass]"
    :size="size"
    :modifier="modifier"
    :label-class="labelClass"
    :content-class="contentClass"
    :background="background"
    :disabled="$attrs.disabled"
  >
    <template #input>
      <input
        :class="[style['cdr-radio__input'], inputClass]"
        type="radio"
        :name="name"
        :value="customValue"
        v-bind="$attrs"
        v-model="radioModel"
      >
    </template>
    <slot />
  </cdr-label-wrapper>
</template>

<style lang="scss" module src="./styles/CdrRadio.module.scss">
</style>
