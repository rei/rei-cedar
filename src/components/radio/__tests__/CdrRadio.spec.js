import { mount } from '../../../../test/vue-jest-style-workaround.js';
import CdrRadio from '../CdrRadio.vue';

describe('CdrRadio', () => {
  describe('with customValue and name prop', () => {
    let wrapper;
    beforeEach(() => {
      wrapper = mount(CdrRadio, {
        propsData: {
          customValue: 'A',
          name: 'testName',
        }
      });
    });

    it('renders correctly', () => {
      expect(wrapper.element).toMatchSnapshot();
    });

    it('is type radio', () => {
      expect(wrapper.find('input').attributes('type')).toBe('radio');
    });

    it('sets name attribute correctly', () => {
      expect(wrapper.find('input').attributes('name')).toBe('testName');
    });
  });

  describe('with custom classes', () => {
    let wrapper;
    beforeEach(() => {
      wrapper = mount(CdrRadio, {
        propsData: {
          labelClass: 'custom-label-class',
          inputClass: 'custom-input-class',
          contentClass: 'custom-content-class',
          customValue: 'A',
          modelValue: 'A',
          name: 'testName',
        },
        slots: {
          default: 'Label Test',
        },
      });
    });

    it('renders correctly', () => {
      expect(wrapper.element).toMatchSnapshot();
    });

    it('adds a custom label class correctly', () => {
      expect(wrapper.find('.custom-label-class').exists()).toBe(true);
    });

    it('adds a custom input class correctly', () => {
      expect(wrapper.find('.custom-input-class').exists()).toBe(true);
    });

    it('adds a custom content class correctly', () => {
      expect(wrapper.find('.custom-content-class').exists()).toBe(true);
    });

    it('evaluates simple checked state correctly', () => {
      expect(wrapper.find('input').element.checked).toBe(true);
    });

    describe('with no modelValue', ()=> {
      beforeEach(()=>{
        wrapper.setProps({ modelValue: '' })
        const rb = wrapper.find('input');
        rb.element.checked = 'true';
        rb.trigger('change')
      })

      it('emits a change event with correct value', () => {
        expect(wrapper.emitted()['update:modelValue'][0][0]).toBe('A');
      });
    })

    describe('when the model value changes', ()=>{
      beforeEach(()=>{
        wrapper.setProps({ modelValue: 'AA' })
      })

      it('evaluates simple not checked state correctly', () => {
        expect(wrapper.find('input').element.checked).toBe(false);
      });
    })

    describe('a complex group with a non-matching modelValue', () => {
      beforeEach(() => {
        wrapper.setProps({
          customValue: { test: 'B', arr: [1, 2, 3] },
          modelValue: { test: 'B' },
        })
      })
      it('evaluates complex group not checked state correctly', () => {
        expect(wrapper.find('input').element.checked).toBe(false);
      });
    })
  });

  // TODO: would need to update CdrRadio.vue to do deepEquality check in order to support complex objects like this.
  // I am not sure if this is something we actually need to support or if it's worth the cost of implementing into CdrRadio?
  // For example, a consumer could bind ID/Key strings to the radio and use them to look up the associated object in their code rather than binding arbitrary objects
  // <cdr-radio v-model="foo">
  // const options = {foo: {test: 'b', arr: [1,2,3]}}, etc.
  // xit('evaluates complex group checked state correctly', () => {
  //   const wrapper = mount(CdrRadio, {
  //     propsData: {
  //       customValue: {test: 'B', arr: [1,2,3]},
  //       name: 'testName',
  //       modelValue: {test: 'B', arr: [1,2,3]},
  //     },
  //   });
  //   expect(wrapper.find('input').element.checked).toBe(true);
  // });
});