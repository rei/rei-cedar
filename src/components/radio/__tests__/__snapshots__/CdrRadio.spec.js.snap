// Vitest Snapshot v1

exports[`CdrRadio > renders an input 1`] = `
<div
  class="cdr-label-wrapper__container cdr-radio"
>
  <label
    class="cdr-label-wrapper cdr-label-wrapper--primary cdr-label-wrapper--medium"
  >
    
    <input
      class="cdr-radio__input"
      name="testName"
      type="radio"
      value="A"
    />
    
    <span
      class="cdr-label-wrapper__figure"
    />
    <span
      class="cdr-label-wrapper__content"
    >
      
      
    </span>
  </label>
</div>
`;

exports[`CdrRadio renders an input 1`] = `
<div
  class="cdr-label-wrapper__container cdr-radio"
>
  <label
    class="cdr-label-wrapper cdr-label-wrapper--primary cdr-label-wrapper--medium"
  >
    
    <input
      class="cdr-radio__input"
      name="testName"
      type="radio"
      value="A"
    />
    
    <span
      class="cdr-label-wrapper__figure"
    />
    <span
      class="cdr-label-wrapper__content"
    >
      
      
    </span>
  </label>
</div>
`;
